### ISSUE TITLE
Disallow new with empty parens

### ISSUE THREAD

**Issue by epoberezkin** [eslint/eslint#6857]
The rule can be implemented as one of the following:
- an option in new-parens rule
- an option in no-extra-parens rule
- a separate rule

**When does this rule warn? Please describe and show example code:**

``` javascript
var p = new Person(); // error
var p = new Person; // correct
var name = new Person().getName(); // optionally correct
```

**Is this rule preventing an error or is it stylistic?**

Stylistic

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

It is generic

**Are you willing to create the rule yourself?**

Yes

**Comment by platinumazure** [comment]
I think the best approach is probably as an option for `new-parens`. Would you mind tweaking your original post to follow our [Propose Rule Change template](http://eslint.org/docs/developer-guide/contributing/rule-changes#proposing-a-rule-change) along those lines? Thanks!

**Comment by eslintbot** [comment]
Hi @epoberezkin, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by epoberezkin** [comment]
**What version of ESLint are you using?**

3.3.0

**What rule do you want to change?**

Rule: new-parens

Possible change:
option `noEmptyParens: true`
option `allowPropertyAccess: true` (maybe this should be another rule?)

**What code should be flagged as incorrect with this change?**

```javascript
var p = new Person(); // error with `noEmptyParens: true`
var p = new Person; // correct with `noEmptyParens: true`
var name = new Person().getName(); // correct with `allowPropertyAccess: true`, error without
```

**What happens when the rule is applied to this code now?**

```
var p = new Person(); // correct
var p = new Person; // error
var name = new Person().getName(); // correct
```

**Comment by btmills** [comment]
What about adding an "always" (default) / "never" mode to `new-parens`?

**Comment by platinumazure** [comment]
@btmills @epoberezkin For this case: `new Person().getName();` Would that automatically not be flagged in "never" mode, giving the developer benefit-of-doubt with the parens probably being necessary? Should there perhaps be an option to require `new (Person.getName)()` for property accesses to avoid ambiguity?

**Comment by epoberezkin** [comment]
I think there should be an option to prohibit immediate property access in the same expression when the instance was constructed. However you do it, it looks not too good (that's an opinion of course, that's why it should be an option :)

**Comment by epoberezkin** [comment]
The variations in the property access, both to access constructor and to access property in constructed instance, probably suggest a separate rule completely - new-property?
Should we maybe focus on the issue at hand, i.e. add modes to new-parens, as @btmills suggested, to either "always" require parens (current mode) or "prohibit" empty parens?

I am ok with anything to be honest, please let me know what you think is more in line with how other rules are organised...

**Comment by platinumazure** [comment]
Well, the only reason I brought it up is because parentheses are definitely required in `var name = new Person().getName();`. But then again, it could also be written as `var name = (new Person).getName();` or similar.

@btmills Since you brought up the elegant solution of always/never and I'm the one throwing a wrench into the works, what do you think? Should we just do always/never in one PR, and then consider another finer-control option later? I'll defer to your advice here.

**Comment by btmills** [comment]
@platinumazure good catch, that's quite the wrench. Let's see... Based on the cases demonstrated below, it looks like we'll need either a tri-state or an option on the "never". I agree with @epoberezkin that immediate property access is pretty ugly, so maybe there's a way to support that too.

```js
/* 1 */ new Class()
/* 2 */ new Class.staticPropertyReturningConstructor()
/* 3 */ new Class().instanceProperty
/* 4 */ new Class().instanceMethod()
/* 5 */ new Class
/* 6 */ new Class.staticPropertyReturningConstructor
/* 7 */ (new Class).instanceProperty
/* 8 */ (new Class).instanceMethod()
```

So what if we started off with "always" (1-4 valid) and "never" (5-8 valid). Then, later, if someone requests an option that makes 3-6 valid, we could add a third state, making it "always"/"never"/"chained" or something like that.

**Comment by gajus** [comment]
This has been in "evaluating" state since August. Can it be assigned a priority?

> So what if we started off with "always" (1-4 valid) and "never" (5-8 valid). Then, later, if someone requests an option that makes 3-6 valid, we could add a third state, making it 

Sounds reasonable approach.

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
